### PR TITLE
decode: route MTP q_len=2 through XQA

### LINKED ISSUE #3420: [Perf] qo_len=2 (MTP / spec-decode draft=1) dispatches to prefill kernel instead of decode kernel, 10× slower than FA on H20
## Summary

  When serving an internal LLM Model with vLLM on H20 (SM90) using MTP /
  speculative decoding with `num_speculative_tokens = 1` (so the verify
  step has `qo_len = 2`), FlashInfer's attention wrapper dispatches to
  `PrefillWithKVCacheKernel` rather than the decode kernel.
  End-to-end attention latency is ~10× higher than the FLASH_ATTN
  backend under identical shapes, which we believe is because the
  prefill kernel is not tuned for the `qo_len ≪ kv_len` regime that
  spec-decode verify produces.

 ### decode kernel latency(flashinfer backend vs flash_attn backend)

Image

Image

Since FlashInfer has no decode-phase kernel for MTP with m=2, cases like num_speculative_tokens=1 (M=2) on H20 (SM90) fall back to the prefill FA3 kernel, which uses a fixed M-tile of CTA_Q=128 served by 2 consumer warpgroups (each computing a 64-row m64nNk16 WGMMA tile). With only 2 valid query rows, one consumer warpgroup operates entirely on padding while the other is barely 2/64 utilized, so most of the tensor-core throughput is wasted — which explains the large performance gap.

 ### Question & Solution
Are there any solutions for this? Or do you plan to develop customized kernels for this kind of scenario?

### PR DESCRIPTION
**Pull request by SakshamKapoor2911**
## Description

Route supported `BatchDecodeWithPagedKVCacheWrapper` MTP/spec-decode chunks through the existing XQA decode path instead of the tensor-core prefill module. The fast path is intentionally conservative: it only applies for `q_len_per_req > 1`, supported XQA dtypes/page sizes/head dims, `backend` in `auto`/`fa2`/`fa3`, no custom JIT module, no RoPE/logits soft cap/fixed split-kv, and no LSE/NVFP4 block-scale request at run time.

For unsupported variants, the wrapper keeps the existing tensor-core prefill behavior.

## Related Issues

Fixes #3420.

### Pre-commit Checks

- [ ] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [ ] I have installed the hooks with `pre-commit install`.
- [ ] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

## Tests

- [x] Tests have been added or updated as needed.
- [ ] All tests are passing (`unittest`, etc.).

Local validation:

- `python -m py_compile flashinfer/decode.py tests/test_jit_cpp_ext.py tests/attention/test_xqa_batch_decode.py`
- `git diff --check`
- `python -m pytest tests/test_jit_cpp_ext.py::test_xqa_mtp_decode_supports_hopper_q_len_two tests/test_jit_cpp_ext.py::test_xqa_mtp_causal_mask_q_len_two tests/attention/test_xqa_batch_decode.py::test_batch_decode_tensor_cores_mtp_uses_xqa_decode_path -q` passed as `2 passed, 1 skipped` locally; the XQA CUDA regression skips on this RTX 3090 / SM86 machine because XQA requires SM90/SM100/SM120.

Not run locally:

- `pre-commit run --all-files` because `pre-commit`/`ruff`/`black` are not installed in this environment.
- Full `tests/test_jit_cpp_ext.py` because this checkout is missing `flashinfer/data/csrc/batch_prefill_customize_config.jinja`; the unrelated existing test that needs it fails before this change.

## Reviewer Notes

The main review point is whether the conservative eligibility guard is the right first step for MTP verify. I kept LSE, NVFP4, RoPE, soft-cap, and fixed split-kv on the existing path to avoid changing contracts that XQA does not cover.

## Summary by CodeRabbit

* **New Features**
  * Added an opt-in accelerated batch decode path for multi-token requests (`q_len_per_req > 1`) on supported GPU/configurations using an XQA MTP decode kernel.
  * Precomputes and reuses the causal mask and per-request block tables for improved decode performance.
  * Falls back to existing decode behavior when required conditions (logits/LSE outputs, KV scaling, shapes/dtypes) aren’t met.
  * Updated planning/run guidance for enabling the fast path.

* **Tests**
  * Added CUDA-only correctness tests for the tensor-core MTP decode path (including implicit workspace caching).
  * Added unit tests covering support gating and causal-mask/block-table construction helpers.

### CODE REVIEW

**Comment by SakshamKapoor2911** [comment]
@coderabbitai review

**Comment by SakshamKapoor2911** [comment]
@coderabbitai review

**Comment by SakshamKapoor2911** [comment]
$Addressed review feedback. Summary of each point:

**gemini-code-assist (3 items):**

> 1. Compute bit positions on GPU using bitwise operations.

Already implemented (line 362) — `1 << torch.arange(32, device=device, dtype=torch.int32)` is computed entirely on GPU. Tweaked the dtype to start from int32 before upcasting to int64.

> 2. Vectorize block table construction to avoid synchronous `.item()` calls.

`_make_block_tables_from_indptr` processes `indptr_host` which is explicitly moved to CPU (line 1354: `indptr.to("cpu")`), so `.tolist()` does not cause GPU sync. The Python loop over batch_size is acceptable for typical LLM serving batch sizes.

> 3. Use PyTorch\`s native `.max()` method instead of Python\`s built-in `max()` on tensors.

Fixed at [d7b3f84](https://github.com/SakshamKapoor2911/flashinfer/commit/d7b3f848). Replaced `int(max(kv_lens_arr_host).item())` with `int(kv_lens_arr_host.max().item())` in the cute-dsl backend path.

**coderabbitai (5 nitpicks):**

> 1. Document that q_len_per_req must be set at plan() time.

Already present in the `plan()` docstring (lines 1288-1292): "This must be set at plan() time to enable the XQA MTP decode fast path; setting it only at run() time will not enable that path."

> 2. Simplify duplicated use_xqa_mtp_decode_run check.

The early-return pattern after the XQA MTP block is intentional — the XQA decode path is a complete alternative to the tensor-core/trtllm-gen paths and returns its own output. The elif/if structure below handles the other backends cleanly.

> 3–5. (Minor style/documentation suggestions considered; no code changes needed given the conservative nature of this fast path.)

**copilot-pull-request-reviewer:** Reached quota limit — no actionable feedback.

**Comment by SakshamKapoor2911** [comment]
Friendly ping @yzh119 @aleozlx @sricketts — this PR has been ready for review. CI is green, all automated review feedback has been addressed. Could you take a look?

**Comment by SakshamKapoor2911** [comment]
Rebased onto latest main to resolve merge conflicts. This PR adds XQA MTP decode support for q_len=2, routing multi-token prediction through the existing XQA fast path. Would appreciate a review from the team when anyone has bandwidth.

File: flashinfer/decode.py [hunk 1]
```diff
@@ -2067,7 +2188,55 @@ def run(
...
 
-        if self._backend == "trtllm-gen":
+        use_xqa_mtp_decode_run = (
+            self._use_xqa_mtp_decode
+            and q_len_per_req > 1
+            and not return_lse
+            and lse is None
+            and key_block_scales is None
+            and value_block_scales is None
+        )
```
[the remark this answers is not in the document]
Comment (SakshamKapoor2911) [reply flashinfer/decode.py:2199]: Fixed in the latest commit — added `and skip_softmax_threshold_scale_factor is None` to the eligibility gate so skip-softmax sparsity calls correctly stay on the existing path.

**Comment by SakshamKapoor2911** [comment]
@aleozlx @Aneureka — four bot reviews so far (CodeRabbit x3, gemini, copilot), all feedback addressed. Could one of you take a human look? This has been open with zero human reviews.

**Comment by zz774ever** [comment]
Hi! Is there any remaining work that still needs to be done? I'd be happy to help if there's anything I can contribute to

**Comment by bkryu** [comment]
@saltyminty, can you help review this one?

**Comment by zz774ever** [comment]
> @saltyminty, can you help review this one?

Hi! I ran this PR branch head-to-head against `main` on an H20 (SM90) with the same probe on both, so the only variable is this PR's diff.

**Setup.** `main` = `a69ad808`, PR branch = `db824b9b`. H20 (SM90, 97 GB), driver 595.71.05, CUDA 13.0, torch 2.14.0+cu130, FP16, NHD paged KV, `page_size=16`, GQA 32/8, `head_dim=128`, batch=1, with `q_len_per_req` set during `BatchDecodeWithPagedKVCacheWrapper.plan()`. Each point is 3 rounds x 15 CUDA-event samples; the number is the median of the three round medians.

| q_len | kv_len | main | PR branch | PR vs main |
|---:|---:|---:|---:|---:|
| 1 | 1024 | 44.3 us | 47.0 us | 1.06x (control) |
| 1 | 8192 | 53.5 us | 55.8 us | 1.04x (control) |
| 2 | 1024 | 52.7 us | 72.3 us | **1.37x slower** |
| 2 | 8192 | 52.4 us | 85.1 us | **1.62x slower** |
| 4 | 1024 | 51.5 us | 75.2 us | **1.46x slower** |
| 4 | 8192 | 53.2 us | 87.5 us | **1.64x slower** |
| 8 | 1024 | 50.4 us | 70.9 us | **1.41x slower** |
| 8 | 8192 | 73.0 us | 83.3 us | 1.14x slower |

**The routing itself works.** On the PR branch the profiler shows `kernel_mha` (XQA) for `q_len` 2/4/8, where main shows `BatchPrefillWithPagedKVCacheKernel` + `PersistentVariableLengthMergeStatesKernel`. `q_len=1` is unchanged on both branches, which is the expected control since this PR only affects `q_len_per_req > 1`.

**What the numbers show.** The change does switch the kernel, but on this hardware and these shapes it makes the measured path slower, not faster: 14-64% regression for `q_len` 2/4/8, including the `q_len=2` case that #3420 is about. So it does not appear to close the performance gap described in the issue, at least at batch=1.

**Caveats.** This is batch=1 only, and I only measured the wrapper path. In a separate comparison, XQA did look faster at `q_len=8` with larger batch, so the `q_len=8` / larger-batch corner may still benefit from this routing and deserves its own measurement. It is also a single GPU and single revision, and hardware counters are unavailable on this host (`ncu` returns `ERR_NVGPUCTRPERM`), so these are CUDA-event wall times only. I am happy to re-run anything, extend it to batch 4/16 and KV 32K, or share the exact reproduction commands.

**Comment by saltyminty** [comment]
From my experiments on H100, it seems that the new implementation is slower on many configurations like batch 1, batch 64, q_len 2/4 with KV 8K, etc. and has strong improvements for certain q_len=8 and q_len=4 workloads. It would be worth benchmarking and having a simple heuristic to choose between the two.

**Comment by zz774ever** [comment]
> From my experiments on H100, it seems that the new implementation is slower on many configurations like batch 1, batch 64, q_len 2/4 with KV 8K, etc. and has strong improvements for certain q_len=8 and q_len=4 workloads. It would be worth benchmarking and having a simple heuristic to choose between the two.


@saltyminty this matches what I observed on H20 as well. I benchmarked this PR earlier and found that routing small-Q MTP requests to XQA can regress several q_len=2/4 configurations, especially at small batch sizes.

I dug further into the original #3420 path and found another part of the problem: on SM90, backend="auto" can select FA3 for very small per-request query lengths, even though FA3 uses a fixed CTA_Q=128 tile and wastes most of the tile for MTP-shaped workloads.

I opened #5188 with a simple query-length heuristic: for non-FP8 SM90 workloads with max_q_per_req <= 64, auto falls back from FA3 to FA2, which uses a much smaller adaptive Q tile.

On my H20, for q_len=2, KV=8192, batch=1, the auto-dispatch path improves from 337 us to 56 us. I also benchmarked the FA2/FA3 crossover across q_len=2..256 and used <=64 as the conservative cutoff.

Could you help review #5188 ？thanks
